### ISSUE TITLE
Bug in width calculation with multibyte characters.

### ISSUE THREAD

**Issue by n2naokun** [leodevbro/vscode-blockman#109]
Note: This text has been translated using DeepL.com.
English
When I use multibyte characters such as Japanese, the frame is not displayed correctly.
Multibyte characters are displayed wider, but all of them seem to be calculated with the same width as alphabets and symbols.

If it is technically possible, we would appreciate it if you could fix this.

Japanese(original)
日本語などのマルチバイト文字を使用した場合枠が正しく表示されません。
マルチバイト文字は幅が広く表示されますが、全てアルファベットや記号と同じ幅で計算されているようです。

技術的に可能でしたら修正していただけると助かります。

![blockman](https://user-images.githubusercontent.com/5172746/230702156-1261bdcc-40b0-4038-9114-f84cca989426.png)

**Comment by leodevbro** [comment]
Hi, I think we can make the situation better, but unfortunately it <a href="https://github.com/microsoft/vscode/issues/118994">does not seem easy</a> to fully fix it, because VS Code extension API does not expose width of each character or location of each character. So, blockman just calculates block width with 1 general character width times the count of characters in the line.

For now, please try this Blockman command:
Press F1 and find:
`Blockman Toggle Try Support Double Width Chars (e.g. Chinese)`

![image](https://user-images.githubusercontent.com/54504880/230712658-165bcdcd-69e4-48a6-ad78-b860f253d682.png)

With this experimental feature, Blockman will try to find Chinese/Japanese characters and it will try to treat each of them as two characters, so it will look something like this. It is still not correct width, but at least the rectangle is bigger, so the text does not overflow and it feels more natural probably.

![image](https://user-images.githubusercontent.com/54504880/230712841-a039f38b-6164-4e87-b2fb-7801e4134c2f.png)

**Comment by n2naokun** [comment]
Thank you very much.
It is now easier to read.

**Comment by ChenZhouUC** [comment]
Thank you a lot for your help. This is enough for me. Still confused for the one point, is this toggle command thing different from the extension settings since I couldn't find any changes on my VSCODE user settings?

**Comment by leodevbro** [comment]
Yeah, it does not change the settings. It just changes the current state in memory. So, probably you have to rerun the command every time you restart VS Code.

**Comment by ChenZhouUC** [comment]
> Yeah, it does not change the settings. It just changes the current state in memory. So, probably you have to rerun the command every time you restart VS Code.

Thank you for your kind reply. Found a way to do this automatically. I would like to put it here in case anyone who has similar concern.

Using VSCode's Tasks:
(ref: https://code.visualstudio.com/docs/editor/tasks)
```{json}
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "blockman",
            "type": "process",
            "command": "${command:blockman.toggleTrySupportDoubleWidthChars}",
            "args": [],
            "group": "none",
            "runOptions": {
                "runOn": "folderOpen"
            }
        }
    ]
}
```

**Comment by ChenZhouUC** [comment]
Still Looking Forward to a formal Setting for this feature. This is indeed a good enough one! I consider wider is better though some outlier cases might occur.

**Comment by leodevbro** [comment]
Thanks for the feedback. I'll try to implement this feature as a formal setting for the next update of Blockman.

**Comment by tats-u** [comment]
How about using [`string-width`](https://github.com/sindresorhus/string-width/issues) or another similar package if you can't calculate the rendered width of each line? It can handle some invisible characters whose width should be zero.

**Comment by leodevbro** [comment]
string-width seems interesting, I'll check it out, thanks. Unfortunately my time is limited these months, so maybe someone else can create a pull request.

**Comment by tats-u** [comment]
> Unfortunately my time is limited these months

I'm not in a hurry at all. After writing that, I became satisfied after discovering and enabling "Blockman Toggle Try Support Double Width Chars (e.g. Chinese)" because it covers most everyday cases.

> so it will look something like this. It is still not correct width,
> 
> ![image](https://user-images.githubusercontent.com/54504880/230712841-a039f38b-6164-4e87-b2fb-7801e4134c2f.png)

I think this depends on the current font. Using a font where the CJK character width is exactly twice that of alphanumeric characters prevents the width misalignment problem described above. The current "Blockman Toggle Try Support Double Width Chars (e.g. Chinese)" meets sufficient quality standards to warrant enabling by default.

Here are some Japanese dedicated fonts that meet such criteria:

- [BIZ UDGothic](https://github.com/googlefonts/morisawa-biz-ud-gothic) (Note: you should install [Japanese supplemental fonts](https://learn.microsoft.com/windows/deployment/windows-missing-fonts?pivots=windows-11) in non-Japanese Windows to add it)
- MS Gothic (ditto in non-Japanese Windows)
- [UDEV Gothic](https://github.com/yuru7/udev-gothic) (JetBrains Mono + BIZ UDGothic) (Note: its author also publishes many other similar Japanese programming fonts)
- [SOROEMONO](https://github.com/qrac/soroemono) (same combination as UDEV Gothic but expands Japanese characters instead of shrinking alphanumeric characters)

In addition, there is [Cascadia Code Next](https://github.com/microsoft/cascadia-code/discussions/762) that also supports Chinese.

<img width="708" height="262" alt="Image" src="https://github.com/user-attachments/assets/54202573-9a7d-47f9-9e34-979f1dfed132" />

This screenshot uses SOROEMONO.
